ksql-rest-app/src/test/java/io/confluent/ksql/rest/server/computation/KafkaConfigStoreTest.java

```diff
@@ -0,0 +1,204 @@
+  }
+
+  @SuppressWarnings("unchecked")
+  private KafkaConsumer<String, KsqlProperties> expectRead(
+      final KafkaConfigStore.KsqlProperties... properties) {
+    final KafkaConsumer<String, KsqlProperties> consumer = mock(KafkaConsumer.class);
+    when(consumer.endOffsets(any()))
+        .thenReturn(ImmutableMap.of(topicPartition, (long)properties.length));
+    when(consumer.position(any())).thenReturn(0L).thenReturn((long) properties.length);
+    final List<ConsumerRecord<String, KafkaConfigStore.KsqlProperties>> records
+        = new LinkedList<>();
+    for (int i = 0; i < properties.length; i++) {
+      records.add(
+          new ConsumerRecord<>(
+              topicName,
+              0,
+              (long) i,
+              "ksql-standalone-configs",
+              properties[i]
+          )
+      );
+    }
+    when(consumer.poll(any())).thenReturn(
+        new ConsumerRecords<>(
+            Collections.singletonMap(
+                topicPartition,
+                records
+            )
+        )
+    );
+    return consumer;
+  }
+
+  private void verifyDrainLog(
+      final KafkaConsumer<String, KsqlProperties> consumer,
+      final InOrder inOrder,
+      final int nmsgs) {
+    inOrder.verify(consumer).assign(topicPartitionAsList);
+    inOrder.verify(consumer).seekToBeginning(topicPartitionAsList);
+    for (int i = 0; i < nmsgs; i++) {
+      inOrder.verify(consumer).poll(any());
+    }
+  }
+
+  @Test
+  public void shouldWriteConfigIfNoConfigWritten() {
+    // Given:
+    final KafkaConsumer<String, KsqlProperties> consumerBefore = expectRead();
+    final KafkaConsumer<String, KsqlProperties> consumerAfter = expectRead(
+        new KafkaConfigStore.KsqlProperties(
+            ksqlConfig.getAllConfigPropsWithSecretsObfuscated()));
+    when(consumerSupplier.get()).thenReturn(consumerBefore).thenReturn(consumerAfter);
+
+    // When:
+    initConfigStore();
+    final KsqlConfig resolvedConfig = configStore.getKsqlConfig();
+
+    // Then:
+    assertThat(resolvedConfig.values(), equalTo(ksqlConfig.values()));
+    assertThat(
+        resolvedConfig.getKsqlStreamConfigProps(),
+        equalTo(ksqlConfig.getKsqlStreamConfigProps()));
+    final InOrder inOrder = Mockito.inOrder(consumerBefore, consumerAfter, producer);
+    verifyDrainLog(consumerBefore, inOrder, 0);
+    @SuppressWarnings("unchecked")
+    final ArgumentCaptor<ProducerRecord<String, KafkaConfigStore.KsqlProperties>> msgCaptor
+        = ArgumentCaptor.forClass(ProducerRecord.class);
+    inOrder.verify(producer).send(msgCaptor.capture());
+    assertThat(
+        msgCaptor.getValue().value().getKsqlProperties(),
+        equalTo(ksqlConfig.getAllConfigPropsWithSecretsObfuscated()));
+    inOrder.verify(producer).flush();
+    verifyDrainLog(consumerAfter, inOrder, 1);
+  }
+
+  @Test
+  public void shouldReadAndMergeExistingConfigIfExists() {
```

Comment:
This test would benefit from being split.

* shouldReadAndMergeExistingConfigIfExists
* shouldNotWriteConfigIfConfigExistsInTopic

and add:

* shouldReturnFirstExistingConfig - i.e. have poll return multiple configs and ensure the first is returned.
* shouldContinueToPollUntilEndOfsetReached - i.e. test that it will continue to poll, and ignore records with wrong key, until end offset. Probably need two tests here: one with a final configs record and one without. 

I also think these tests would benefit from a restructor to make them more readable.  If you add a 'firstConsumer' and 'secondConsumer' mocks and set up the `consumerSupplier` in `setUp`, then tests can look more like:

```java
// Given:
givenExistingPersistedConfig(fistConfig, secondConfig);

// When:
final KsqlConfig resolvedConfig = configStore.getKsqlConfig();

// Then:
assertThat(resolvedConfig, is(firstConfig);
```

Reply:
> shouldReturnFirstExistingConfig - i.e. have poll return multiple configs and ensure the first is returned.

I don't know that it makes sense to have a test case for this. It's valid to read any of the values with the given key (we're using a compacted topic). The expectation is that all servers write out the same config.

Reply:
I'll just make it a non-compacted topic and add that test. We get stronger guarantees and the cost is tiny.